File: android/sdk/src/main/kotlin/dev/glassfalcon/core/MissionEngine.kt
// GlassFalcon, authored and owned by FalconTechnix.
// Copyright (C) 2026 FalconTechnix.
// SPDX-License-Identifier: GPL-3.0-or-later
//
// Free software released by FalconTechnix under the GNU GPL v3 or later.
// See LICENSE at the repository root or <https://www.gnu.org/licenses/>.

package dev.glassfalcon.core

import kotlinx.coroutines.*
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlin.math.*

class MissionEngine(
    private val duml: DumlConnection,
    private val scope: CoroutineScope,
) {
    private val _status = MutableStateFlow(MissionStatus())
    val status: StateFlow<MissionStatus> = _status

    // Navigation tuning constants
    private val YAW_KP    = 0.8f   // yaw proportional gain
    private val ALT_KP    = 0.4f   // altitude P gain
    private val POS_KP    = 0.6f   // horizontal position P gain
    private val MAX_SPEED = 8f     // m/s target horizontal speed
    private val YAW_THRESHOLD = 15f  // degrees, aligned enough to start moving
    private val WP_RADIUS_M   = 3f   // metres, close enough to waypoint

    private var plan: MissionPlan? = null
    private var droneState: DroneState = DroneState()
    private var navJob: Job? = null

    // Freshness of droneState. The nav loops are open-loop P-controllers driving throttle
    // and pitch from telemetry; if telemetry stalls (link loss) the last value freezes and
    // the controller would keep commanding a stale error, e.g. hold full climb. Track the
    // arrival time and bail to neutral sticks the moment it goes stale.
    @Volatile private var lastStateMs = 0L
    private val STALE_MS = 1500L

    fun attachDroneState(state: DroneState) {
        droneState = state
        lastStateMs = System.currentTimeMillis()
    }

    private fun telemetryFresh() =
        lastStateMs != 0L && System.currentTimeMillis() - lastStateMs < STALE_MS

    // ── Start / Stop ─────────────────────────────────────────────────────────

    fun start(p: MissionPlan, startAlt: Float = p.waypoints.firstOrNull()?.altM ?: 50f) {
        if (navJob?.isActive == true) return
        plan = p
        _status.value = MissionStatus(
            state = MissionState.TAKEOFF,
            totalWaypoints = p.waypoints.size,
            log = listOf("Mission start: ${p.name}, ${p.waypoints.size} waypoints"),
        )
        navJob = scope.launch { runMission(p, startAlt) }
    }

    fun abort() {
        navJob?.cancel()
        sendSticks(0f, 0f, 0f, 0f)   // stop
        _status.value = _status.value.copy(state = MissionState.ABORTED)
        log("Mission ABORTED")
    }

    // ── Main mission coroutine ────────────────────────────────────────────────

    private suspend fun runMission(p: MissionPlan, takeoffAlt: Float) {
        // ── 1. Climb to altitude ──────────────────────────────────────────────
        setState(MissionState.TAKEOFF)
        log("Climbing to ${takeoffAlt.toInt()} m…")
        climbToAlt(takeoffAlt)

        // ── 2. Fly each waypoint ──────────────────────────────────────────────
        var photos = 0
        var distTravelled = 0f
        var prevLat = droneState.lat
        var prevLon = droneState.lon

        for ((idx, wp) in p.waypoints.withIndex()) {
            if (!isActive()) break

            setState(MissionState.FLYING)
            _status.value = _status.value.copy(
                currentWpIdx   = idx,
                photosTaken    = photos,
                distanceTravelledM = distTravelled,
                etaSeconds     = estimateEta(p, idx),
            )
            log("→ WP $idx ${wp.label}  (${wp.lat.fmt()}, ${wp.lon.fmt()}) ${wp.altM.toInt()}m")

            // Set gimbal pitch for this leg
            setGimbalPitch(wp.gimbalPitch)

            // Fly to waypoint
            flyToWp(wp)
            if (!isActive()) break

            // Accumulate distance
            distTravelled += MissionPlanner.haversine(prevLat, prevLon, droneState.lat, droneState.lon).toFloat()
            prevLat = droneState.lat; prevLon = droneState.lon

            // Capture
            if (wp.capturePhoto) {
                setState(MissionState.CAPTURING)
                log("📷 Photo $photos at WP $idx")
                if (wp.waitMs > 0) delay(wp.waitMs)
                duml.sendCam(Camera.capturePhoto())
                delay(800)   // shutter + buffer
                photos++
            }
        }

        _status.value = _status.value.copy(photosTaken = photos)

        if (!isActive()) return

        // ── 3. Return to home ─────────────────────────────────────────────────
        log("Mission complete ($photos photos). Returning home…")
        setState(MissionState.RETURN_HOME)
        // Signal the FC to do RTH via DUML. Use the single canonical FlyC.returnHome()
        // builder so there is exactly ONE RTH encoding in the codebase (there used to be
        // a second, conflicting 0x03/0x24 definition here).
        duml.send(FlyC.returnHome())
        delay(2000)

        setState(MissionState.COMPLETE)
        log("✓ Mission complete.")
    }

    // ── Navigation primitives ─────────────────────────────────────────────────

    private suspend fun climbToAlt(targetAlt: Float) {
        val deadline = System.currentTimeMillis() + 60_000L
        while (isActive() && System.currentTimeMillis() < deadline) {
            if (!telemetryFresh()) {          // link lost mid-climb, stop commanding climb
                sendSticks(0f, 0f, 0f, 0f)
                log("Telemetry stale during climb, holding, aborting mission")
                abort(); return
            }
            val err = targetAlt - droneState.altRel
            if (abs(err) < 2f) { sendSticks(0f, 0f, 0f, 0f); return }
            val thr = (err * ALT_KP).coerceIn(-0.7f, 0.7f)
            sendSticks(0f, 0f, thr, 0f)
            delay(50)
        }
    }

    private suspend fun flyToWp(wp: Waypoint) {
        val deadline = System.currentTimeMillis() + 120_000L
        while (isActive() && System.currentTimeMillis() < deadline) {
            if (!telemetryFresh()) {          // link lost mid-leg, stop, don't fly blind
                sendSticks(0f, 0f, 0f, 0f)
                log("Telemetry stale en route, holding, aborting mission")
                abort(); return
            }
            val distM = MissionPlanner.haversine(droneState.lat, droneState.lon, wp.lat, wp.lon).toFloat()
            if (distM < WP_RADIUS_M) {
                sendSticks(0f, 0f, 0f, 0f)
                setState(MissionState.AT_WAYPOINT)
                return
            }

            // Where the nose points and which way the drone actually needs to MOVE are two
            // separate things once yawOverrideDeg is in play (a POI/orbit/QuickShot leg wants
            // the camera locked on its subject while flying tangentially past it, not facing
            // the direction of travel). travelBearing is always the true world-frame heading
            // toward the target lat/lon; yawTarget is what the nose should point at.
            val travelBearing = MissionPlanner.bearing(droneState.lat, droneState.lon, wp.lat, wp.lon)
            val yawTarget = wp.yawOverrideDeg?.toDouble() ?: travelBearing
            val headingErr = MissionPlanner.normalizeAngle(yawTarget - droneState.yaw.toDouble()).toFloat()
            val altErr   = wp.altM - droneState.altRel

            val yawOut  = (headingErr / 90f * YAW_KP).coerceIn(-1f, 1f)
            val thr     = (altErr * ALT_KP).coerceIn(-0.5f, 0.5f)

            // Decompose the movement toward travelBearing into THIS frame's forward/lateral
            // stick axes (relative to current yaw), this is what lets an orbit leg move
            // tangentially around its circle while the nose stays locked on the center: at a
            // 90° nose-to-travel offset (the orbit case) this naturally resolves to pure roll,
            // zero pitch, instead of assuming travel is always straight ahead of the nose.
            val travelAngleDeg = MissionPlanner.normalizeAngle(travelBearing - droneState.yaw.toDouble())
            val travelAngleRad = Math.toRadians(travelAngleDeg)
            val speedFrac = (distM / 20f).coerceAtMost(1f)   // slow down as we approach
            // Guard against flying backward while a plain waypoint leg is still mid-turn (no
            // yawOverride, so nose and travel direction should coincide once yaw catches up;
            // until then don't drive off in whatever direction the nose happens to face).
            val moveScale = if (wp.yawOverrideDeg == null && abs(travelAngleDeg) > YAW_THRESHOLD.toDouble() * 2) 0f else 1f
            val pitchOut = (cos(travelAngleRad) * speedFrac * POS_KP * moveScale).toFloat()
            val rollOut  = (sin(travelAngleRad) * speedFrac * POS_KP * moveScale).toFloat()

            sendSticks(roll = rollOut, pitch = pitchOut, throttle = thr, yaw = yawOut)
            delay(50)
        }
        log("WP timeout, continuing to next")
    }

    // ── Helpers ───────────────────────────────────────────────────────────────

    private fun sendSticks(roll: Float, pitch: Float, throttle: Float, yaw: Float) {
        duml.send(FlyC.joystick(roll, pitch, throttle, yaw))
    }

    private fun setGimbalPitch(pitch: Float) {
        duml.sendGimbal(Gimbal.absAngle(pitch, 0f, 0f, 10))
    }

    private fun setState(s: MissionState) {
        _status.value = _status.value.copy(state = s)
    }

    private fun log(msg: String) {
        val cur = (_status.value.log + msg).takeLast(100)
        _status.value = _status.value.copy(log = cur)
    }

    private fun isActive() = navJob?.isActive == true

    private fun estimateEta(p: MissionPlan, currentIdx: Int): Int {
        if (currentIdx >= p.waypoints.size - 1) return 0
        var dist = 0.0
        val wps = p.waypoints
        for (i in currentIdx until wps.size - 1) {
            dist += MissionPlanner.haversine(wps[i].lat, wps[i].lon, wps[i+1].lat, wps[i+1].lon)
        }
        val speed = p.survey?.speedMs?.toDouble() ?: 4.0
        return (dist / speed).toInt()
    }

    private fun Double.fmt() = "%.5f".format(this)
}

File: sdk/python/src/glassfalcon/mission_engine.py
# GlassFalcon, authored and owned by FalconTechnix.
# Copyright (C) 2026 FalconTechnix.
# SPDX-License-Identifier: GPL-3.0-or-later
#
# Free software released by FalconTechnix under the GNU GPL v3 or later.
# See LICENSE at the repository root or <https://www.gnu.org/licenses/>.

"""
Desktop mission engine, executes MissionPlan waypoints via DUML virtual RC sticks.

P-controller navigation: yaw toward bearing first, pitch forward when aligned,
throttle for altitude hold. Same approach as Android MissionEngine.kt.
Runs in a daemon thread so Qt UI is never blocked.
"""

from __future__ import annotations

import math
import threading
import time
from dataclasses import dataclass, field
from enum import Enum
from typing import Callable, Optional

from .mission_planner import MissionPlan, Waypoint, haversine, bearing
from .duml_cmds import (
    flyc_joystick, flyc_motor_ctrl, flyc_emergency_stop,
    camera_capture_photo, gimbal_abs_angle,
    FC, CAM, GIMB, PC, REQ,
)

# ── P-controller constants ─────────────────────────────────────────────────────
YAW_KP       = 0.8    # proportional gain yaw
ALT_KP       = 0.4    # proportional gain altitude
POS_KP       = 0.6    # proportional gain forward pitch
WP_RADIUS_M  = 4.0    # waypoint capture radius
YAW_THRESH   = 12.0   # degrees, must be aligned before pitching forward
MAX_PITCH    = 0.6    # ±1.0 stick scale
MAX_THROTTLE = 0.5
LOOP_HZ      = 10     # navigation loop rate


class MissionState(Enum):
    IDLE      = "IDLE"
    ARMING    = "ARMING"
    TAKEOFF   = "TAKEOFF"
    FLYING    = "FLYING"
    CAPTURING = "CAPTURING"
    RTH       = "RTH"
    LANDING   = "LANDING"
    COMPLETE  = "COMPLETE"
    ABORTED   = "ABORTED"


@dataclass
class MissionStatus:
    state:            MissionState = MissionState.IDLE
    wp_idx:           int = 0
    total_wps:        int = 0
    photos_taken:     int = 0
    distance_m:       float = 0.0
    log:              list[str] = field(default_factory=list)


class MissionEngine:
    """Thread-based mission executor. Call start(plan) to begin."""

    def __init__(self, duml):
        self._duml = duml
        self._state = MissionState.IDLE
        self._status = MissionStatus()
        self._plan: Optional[MissionPlan] = None
        self._abort = threading.Event()
        self._thread: Optional[threading.Thread] = None
        self._on_status: list[Callable[[MissionStatus], None]] = []
        # Latest drone state injected externally
        self._drone_lat  = 0.0
        self._drone_lon  = 0.0
        self._drone_alt  = 0.0
        self._drone_yaw  = 0.0
        self._drone_batt = 100

    # ── Public API ─────────────────────────────────────────────────────────────

    def on_status(self, cb: Callable[[MissionStatus], None]):
        self._on_status.append(cb)

    def update_drone_state(self, lat: float, lon: float, alt: float,
                           yaw: float, batt: int):
        self._drone_lat  = lat
        self._drone_lon  = lon
        self._drone_alt  = alt
        self._drone_yaw  = yaw
        self._drone_batt = batt

    def start(self, plan: MissionPlan):
        if self._thread and self._thread.is_alive():
            self.abort()
        self._plan = plan
        self._abort.clear()
        self._thread = threading.Thread(target=self._run, daemon=True)
        self._thread.start()

    def abort(self):
        self._abort.set()
        self._send_joystick(0, 0, 0, 0)
        self._set_state(MissionState.ABORTED)
        if self._thread:
            self._thread.join(timeout=3)

    # ── Navigation loop ────────────────────────────────────────────────────────

    def _run(self):
        plan = self._plan
        self._status = MissionStatus(total_wps=len(plan.waypoints))
        self._log(f"Mission '{plan.name}' started, {len(plan.waypoints)} waypoints")

        self._set_state(MissionState.ARMING)
        self._send(FC, *flyc_motor_ctrl(True))
        time.sleep(3.0)
        if self._abort.is_set():
            return

        self._set_state(MissionState.TAKEOFF)
        target_alt = plan.waypoints[0].alt_m if plan.waypoints else 30.0
        self._climb_to_alt(target_alt)

        for i, wp in enumerate(plan.waypoints):
            if self._abort.is_set():
                break
            self._status.wp_idx = i
            self._set_state(MissionState.FLYING)
            self._log(f"→ WP {i + 1}/{len(plan.waypoints)}, {wp.label}")
            self._fly_to_wp(wp)

            if self._abort.is_set():
                break

            if wp.capture:
                self._set_state(MissionState.CAPTURING)
                # Point gimbal
                self._send(GIMB, *gimbal_abs_angle(wp.gimbal_pitch, 0.0, 0.0))
                time.sleep(0.6)
                self._send(CAM, *camera_capture_photo())
                self._status.photos_taken += 1
                self._log(f"  Photo {self._status.photos_taken} captured")
                time.sleep(1.5)

        if not self._abort.is_set():
            self._set_state(MissionState.RTH)
            self._log("Mission complete, return to home")
            self._send(FC, 0x03, 0x24, bytes([0x01]))  # RTH cmd_id=0x24
            time.sleep(30)  # wait for RTH to complete (drone handles it)
            self._set_state(MissionState.COMPLETE)
            self._log("Done.")
        else:
            self._send_joystick(0, 0, 0, 0)
            self._log("Mission aborted")

    def _fly_to_wp(self, wp: Waypoint):
        while not self._abort.is_set():
            dist = haversine(self._drone_lat, self._drone_lon, wp.lat, wp.lon)
            if dist < WP_RADIUS_M:
                self._status.distance_m += dist
                return

            # Altitude control
            alt_err = wp.alt_m - self._drone_alt
            throttle = max(-MAX_THROTTLE, min(MAX_THROTTLE, alt_err * ALT_KP))

            # Bearing control
            target_bear = bearing(self._drone_lat, self._drone_lon, wp.lat, wp.lon)
            yaw_err = self._normalize_angle(target_bear - self._drone_yaw)
            yaw = max(-1.0, min(1.0, yaw_err * YAW_KP / 180.0))

            # Only pitch forward once yaw is roughly aligned
            if abs(yaw_err) < YAW_THRESH:
                pitch = min(MAX_PITCH, dist * POS_KP / 100.0)
            else:
                pitch = 0.0

            self._send_joystick(0.0, pitch, throttle, yaw)
            self._notify()
            time.sleep(1.0 / LOOP_HZ)

        self._send_joystick(0, 0, 0, 0)

    def _climb_to_alt(self, target_m: float):
        for _ in range(int(target_m / 0.5 / (1.0 / LOOP_HZ))):
            if self._abort.is_set():
                return
            alt_err = target_m - self._drone_alt
            if abs(alt_err) < 1.0:
                break
            throttle = max(0.05, min(MAX_THROTTLE, alt_err * ALT_KP))
            self._send_joystick(0, 0, throttle, 0)
            time.sleep(1.0 / LOOP_HZ)
        self._send_joystick(0, 0, 0, 0)

    # ── Helpers ────────────────────────────────────────────────────────────────

    @staticmethod
    def _normalize_angle(deg: float) -> float:
        while deg >  180: deg -= 360
        while deg < -180: deg += 360
        return deg

    def _send(self, dst: int, cmd_set: int, cmd_id: int, payload: bytes = b""):
        try:
            self._duml.send(dst=dst, src=PC, cmd_type=REQ,
                            cmd_set=cmd_set, cmd_id=cmd_id, payload=payload)
        except Exception:
            pass

    def _send_joystick(self, roll: float, pitch: float,
                       throttle: float, yaw: float):
        self._send(FC, *flyc_joystick(roll, pitch, throttle, yaw))

    def _set_state(self, state: MissionState):
        self._state = state
        self._status.state = state
        self._notify()

    def _log(self, msg: str):
        ts = time.strftime("%H:%M:%S")
        self._status.log.append(f"{ts}  {msg}")
        self._status.log = self._status.log[-100:]
        self._notify()

    def _notify(self):
        import copy
        st = copy.copy(self._status)
        for cb in self._on_status:
            try:
                cb(st)
            except Exception:
                pass
